[GLIB] Use timerfd to improve timer resolution

GeorgesStavracas opened the pull request:
#### 3a7d28d2021a23bd6607feedb9407638fc14e88f
<pre>
[GLIB] Use timerfd to improve timer resolution
<a href="https://bugs.webkit.org/show_bug.cgi?id=303921">https://bugs.webkit.org/show_bug.cgi?id=303921</a>

Reviewed by Adrian Perez de Castro.

WebKit timers in GLib have a big lateness range (up or more than 1ms).
We can have better timers by using timerfd in a few selected places.

Make RunLoopSource use a timerfd when starting RunLoop::TimerBase
instances. The timerfd itself is created as late as possible, and
only when the timer has an interval set. That&apos;s because some timers
are created but rarely started, so we can save a few fds by delaying
the timerfd creation.

Some measurements. Before this patch:

```
WebPageProxy::GeneratePageLoadTimingTimer lateness: 59µs
MainThreadSharedTimer::timer lateness: 613µs
MainThreadSharedTimer::timer lateness: 92µs
MainThreadSharedTimer::timer lateness: 500µs
BitmapTexturePool::ReleaseUnusedTexturesTimer lateness: 2µs
BitmapTexturePool::ReleaseUnusedTexturesTimer lateness: 867µs
MainThreadSharedTimer::timer lateness: 605µs
BitmapTexturePool::ReleaseUnusedTexturesTimer lateness: 527µs
MainThreadSharedTimer::timer lateness: 194µs
MainThreadSharedTimer::timer lateness: 487µs
BitmapTexturePool::ReleaseUnusedTexturesTimer lateness: 612µs
MainThreadSharedTimer::timer lateness: 687µs
BitmapTexturePool::ReleaseUnusedTexturesTimer lateness: 423µs
MainThreadSharedTimer::timer lateness: 305µs
MainThreadSharedTimer::timer lateness: 161µs
MainThreadSharedTimer::timer lateness: 49µs
BitmapTexturePool::ReleaseUnusedTexturesTimer lateness: 41µs
MainThreadSharedTimer::timer lateness: 431µs
MainThreadSharedTimer::timer lateness: 883µs
JSRunLoopTimer::Manager::PerVMData::Timer lateness: 738µs
BitmapTexturePool::ReleaseUnusedTexturesTimer lateness: 496µs
JSRunLoopTimer::Manager::PerVMData::Timer lateness: 255µs
JSRunLoopTimer::Manager::PerVMData::Timer lateness: 124µs
```

After this patch:

```
WebPageProxy::GeneratePageLoadTimingTimer lateness: 15µs
BitmapTexturePool::ReleaseUnusedTexturesTimer lateness: 10µs
BitmapTexturePool::ReleaseUnusedTexturesTimer lateness: 6µs
MainThreadSharedTimer::timer lateness: 9µs
MainThreadSharedTimer::timer lateness: 10µs
BitmapTexturePool::ReleaseUnusedTexturesTimer lateness: 9µs
MainThreadSharedTimer::timer lateness: 7µs
MainThreadSharedTimer::timer lateness: 416µs
BitmapTexturePool::ReleaseUnusedTexturesTimer lateness: 10µs
MainThreadSharedTimer::timer lateness: 10µs
BitmapTexturePool::ReleaseUnusedTexturesTimer lateness: 12µs
MainThreadSharedTimer::timer lateness: 8µs
MainThreadSharedTimer::timer lateness: 7µs
BitmapTexturePool::ReleaseUnusedTexturesTimer lateness: 10µs
MainThreadSharedTimer::timer lateness: 19µs
MainThreadSharedTimer::timer lateness: 20µs
JSRunLoopTimer::Manager::PerVMData::Timer lateness: 35µs
BitmapTexturePool::ReleaseUnusedTexturesTimer lateness: 15µs
JSRunLoopTimer::Manager::PerVMData::Timer lateness: 24µs
JSRunLoopTimer::Manager::PerVMData::Timer lateness: 12µs
PrivateClickMeasurementManager::FirePendingAttributionRequestsTimer lateness: 24µs
HysteresisActivity::Timer lateness: 14µs
```

It doesn&apos;t seem to affect MotionMark results.

* Source/WTF/wtf/glib/RunLoopGLib.cpp:
(WTF::RunLoop::RunLoop):
(WTF::RunLoop::TimerBase::TimerBase):
(WTF::RunLoop::TimerBase::start):
* Source/cmake/OptionsCommon.cmake:

Canonical link: <a href="https://commits.webkit.org/304553@main">https://commits.webkit.org/304553@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1a5a11853de2db2b3dd0632364e0b882a6abca96

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/135686 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/8060 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/46978 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/143402 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/87340 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/8705 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/7905 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/103695 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/71656 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/138632 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/6276 "Passed tests") | [⏳ 🧪 api-mac](https://ews-build.webkit.org/#/builders/API-Tests-macOS-EWS "Waiting to run tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/84569 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/6041 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/3656 "Passed tests") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/4008 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/127677 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/115267 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/39816 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/146150 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/134170 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/7740 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/40384 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/112056 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/7777 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/6507 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/112431 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/28579 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/5905 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/117929 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/61702 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/7793 "Built successfully") | [⏳ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/macOS-Sonoma-Release-WK2-Intel-Tests-EWS "Waiting to run tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/167001 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/7537 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/71346 "Built successfully") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/43594 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/7757 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/7634 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mcatanzaro commented on Source/WTF/wtf/glib/RunLoopGLib.cpp:
```diff
@@ -282,6 +330,15 @@ void RunLoop::TimerBase::updateReadyTime()
+    // timers are created but may never be triggered.
+    auto& runLoopSource = *reinterpret_cast<RunLoopSource*>(m_source.get());
+    if (m_interval && runLoopSource.timerFd < 0) {
+        runLoopSource.timerFd = timerfd_create(CLOCK_MONOTONIC, TFD_NONBLOCK | TFD_CLOEXEC);
```
If we're worried about nanoseconds, maybe we want `CLOCK_MONOTONIC_RAW` to avoid frequency drift when the system time is adjusted? See `clock_gettime(2)` for the difference.

GeorgesStavracas replied:
The order of magnitude that I was worried about is microseconds, nanoseconds is a bit of an overkill. This patch is inspired by Mutter, which uses CLOCK_MONOTONIC, so I don't think the clock drift is a problem in practice? Also, CLOCK_MONOTONIC is standard Unix but CLOCK_MONOTONIC_RAW is Linux-specific, is that not going to be a problem?

aperezdc replied:
NetBSD and FreeBSD have the `timerfd` functions, but do _not_ have `CLOCK_MONOTONIC_RAW`, AFAICS.

For another data point: the “choreographer” component of the Android compositor also uses `CLOCK_MONOTONIC`, even when `CLOCK_MONOTONIC_RAW` is also available and present in the NDK headers.

mcatanzaro replied:
The question is: do you want the time to actually represent one microsecond in the real world (in which case, you really do have to use `CLOCK_MONOTONIC_RAW`) or are you OK with that microsecond being slightly shorter or longer than a real microsecond when the clock is trying to gradually adjust to a new time (in which case, `CLOCK_MONOTONIC` is perfectly fine).

GLib's monotonic time is notably also `CLOCK_MONOTONIC`.

I'm not too worried about this. Just something to think about.

GeorgesStavracas replied:
I think it is okay (and maybe even more appropriate) to use CLOCK_MONOTONIC. I can't think of an excellent reason to go out of the way to use CLOCK_MONOTONIC_RAW. We just don't need the hardware-bounded high precision of CLOCK_MONOTONIC_RAW.

aperezdc commented on Source/WTF/wtf/glib/RunLoopGLib.cpp:
```diff
@@ -35,21 +36,68 @@
+        *timeout = -1;
+
+        if (runLoopSource.timerFd > -1) {
+            struct itimerspec timerFdSpec = { { 0, 0 }, { 0, 0 } };
```
This could be a plain assignment to `= { }` to have the compiler zero-initialize the `struct`:


```suggestion
            struct itimerspec timerFdSpec = { };
```

aperezdc commented on Source/WTF/wtf/glib/RunLoopGLib.cpp:
```diff
@@ -58,7 +106,18 @@ GSourceFuncs RunLoop::s_runLoopSourceFunctions = {
+#if HAVE(TIMERFD)
+    [](GSource* source) -> void {
+        auto& runLoopSource = *reinterpret_cast<RunLoopSource*>(source);
+        if (runLoopSource.timerFd > -1) {
+            close(runLoopSource.timerFd);
+            runLoopSource.timerFd = -1;
+        }
```
This would be equivalent to `g_clear_fd`:

```suggestion
        g_clear_fd(&runLoopSource.timerFd, nullptr);
```

GeorgesStavracas replied:
g_clear_fd was added in GLib 2.76 whereas WebKit depends on GLib 2.70, we'd need a version bump to use it

aperezdc replied:
Let's then leave this part of the patch as-is.

aperezdc commented on Source/WTF/wtf/glib/RunLoopGLib.cpp:
```diff
@@ -282,6 +347,17 @@ void RunLoop::TimerBase::updateReadyTime()
+    if (m_interval && runLoopSource.timerFd < 0) {
+        runLoopSource.timerFd = timerfd_create(CLOCK_MONOTONIC, TFD_NONBLOCK | TFD_CLOEXEC);
+        if (runLoopSource.timerFd > -1)
+            g_source_add_unix_fd(m_source.get(), runLoopSource.timerFd, G_IO_IN);
```
I think we might want to `LOG_ERROR()` here in case creating the timerfd failed, as it could be useful for debugging. If we are running out of file descriptors at least we would know what's going on instead of silently falling back to “less good” timing.


```suggestion
        if (runLoopSource.timerFd > -1) [[likely]]
            g_source_add_unix_fd(m_source.get(), runLoopSource.timerFd, G_IO_IN);
        else
            LOG_ERROR("Could not create timerfd: %s", safeStrerror(errno).data());
```

aperezdc reviewed (approved):
LGTM, with a couple of comments for possible improvements to apply before landing.

nikolaszimmermann commented on Source/WTF/wtf/glib/RunLoopGLib.cpp:
```diff
@@ -35,21 +36,68 @@
+#if HAVE(TIMERFD) && USE(SYSPROF_CAPTURE)
+        if (readyTime > 0) {
+            gint64 lateness = g_get_monotonic_time() - readyTime;
+            WTFEmitSignpost(source, RunLoopSourceDispatch, "[%s] lateness=%ldµs", g_source_get_name(source), lateness);
```
I'm worried that we're swamped with marks in sysprof. How about adding an env var to toggle this at runtime, defaulting to false. This we we can utilize the mechanism if needed, and avoid adding many new marks, if not necessary.

GeorgesStavracas replied:
Added an env var to control this!

GeorgesStavracas committed:
[GLIB] Use timerfd to improve timer resolution
https://bugs.webkit.org/show_bug.cgi?id=303921

Reviewed by Adrian Perez de Castro.

WebKit timers in GLib have a big lateness range (up or more than 1ms).
We can have better timers by using timerfd in a few selected places.

Make RunLoopSource use a timerfd when starting RunLoop::TimerBase
instances. The timerfd itself is created as late as possible, and
only when the timer has an interval set. That's because some timers
are created but rarely started, so we can save a few fds by delaying
the timerfd creation.

Some measurements. Before this patch:

```
WebPageProxy::GeneratePageLoadTimingTimer lateness: 59µs
MainThreadSharedTimer::timer lateness: 613µs
MainThreadSharedTimer::timer lateness: 92µs
MainThreadSharedTimer::timer lateness: 500µs
BitmapTexturePool::ReleaseUnusedTexturesTimer lateness: 2µs
BitmapTexturePool::ReleaseUnusedTexturesTimer lateness: 867µs
MainThreadSharedTimer::timer lateness: 605µs
BitmapTexturePool::ReleaseUnusedTexturesTimer lateness: 527µs
MainThreadSharedTimer::timer lateness: 194µs
MainThreadSharedTimer::timer lateness: 487µs
BitmapTexturePool::ReleaseUnusedTexturesTimer lateness: 612µs
MainThreadSharedTimer::timer lateness: 687µs
BitmapTexturePool::ReleaseUnusedTexturesTimer lateness: 423µs
MainThreadSharedTimer::timer lateness: 305µs
MainThreadSharedTimer::timer lateness: 161µs
MainThreadSharedTimer::timer lateness: 49µs
BitmapTexturePool::ReleaseUnusedTexturesTimer lateness: 41µs
MainThreadSharedTimer::timer lateness: 431µs
MainThreadSharedTimer::timer lateness: 883µs
JSRunLoopTimer::Manager::PerVMData::Timer lateness: 738µs
BitmapTexturePool::ReleaseUnusedTexturesTimer lateness: 496µs
JSRunLoopTimer::Manager::PerVMData::Timer lateness: 255µs
JSRunLoopTimer::Manager::PerVMData::Timer lateness: 124µs
```

After this patch:

```
WebPageProxy::GeneratePageLoadTimingTimer lateness: 15µs
BitmapTexturePool::ReleaseUnusedTexturesTimer lateness: 10µs
BitmapTexturePool::ReleaseUnusedTexturesTimer lateness: 6µs
MainThreadSharedTimer::timer lateness: 9µs
MainThreadSharedTimer::timer lateness: 10µs
BitmapTexturePool::ReleaseUnusedTexturesTimer lateness: 9µs
MainThreadSharedTimer::timer lateness: 7µs
MainThreadSharedTimer::timer lateness: 416µs
BitmapTexturePool::ReleaseUnusedTexturesTimer lateness: 10µs
MainThreadSharedTimer::timer lateness: 10µs
BitmapTexturePool::ReleaseUnusedTexturesTimer lateness: 12µs
MainThreadSharedTimer::timer lateness: 8µs
MainThreadSharedTimer::timer lateness: 7µs
BitmapTexturePool::ReleaseUnusedTexturesTimer lateness: 10µs
MainThreadSharedTimer::timer lateness: 19µs
MainThreadSharedTimer::timer lateness: 20µs
JSRunLoopTimer::Manager::PerVMData::Timer lateness: 35µs
BitmapTexturePool::ReleaseUnusedTexturesTimer lateness: 15µs
JSRunLoopTimer::Manager::PerVMData::Timer lateness: 24µs
JSRunLoopTimer::Manager::PerVMData::Timer lateness: 12µs
PrivateClickMeasurementManager::FirePendingAttributionRequestsTimer lateness: 24µs
HysteresisActivity::Timer lateness: 14µs
```

It doesn't seem to affect MotionMark results.

* Source/WTF/wtf/glib/RunLoopGLib.cpp:
(WTF::RunLoop::RunLoop):
(WTF::RunLoop::TimerBase::TimerBase):
(WTF::RunLoop::TimerBase::start):
* Source/cmake/OptionsCommon.cmake:

Canonical link: https://commits.webkit.org/304553@main

carlosgcampos commented on Source/WTF/wtf/glib/RunLoopGLib.cpp:
```diff
@@ -35,21 +36,78 @@
     GSource source;
     RunLoop* runLoop;
+#if HAVE(TIMERFD)
+    int timerFd;
```
I'm sorry to be late here, but I think we should use UnixFileDescriptor.

carlosgcampos replied:
Oh, I see this is allocated by GLib...
